common/iterator/sorted/IntersectForwardableIterator.java

```diff
@@ -112,12 +153,15 @@ public T peek() {
     public T next() {
         if (!hasNext()) throw new NoSuchElementException();
         state = State.EMPTY;
+        intersectionValues.add(candidate);
         return candidate;
     }
 
     @Override
     public void forward(T target) {
         iterators.forEach(iterator -> iterator.forward(target));
+        intersectionIterators.clear();
+        intersectionValues.clear();
         state = State.INIT;
     }
```

Comment:
This smells like trouble to me.
This next part is under an assumption that the order of the extra fields is not guaranteed, that is if an iterator is sorted by the first field, we could get `[(1,2), (1,5)]` or `[(1,5), (1,2)]`.
Say we have two iterators A `[..., (1,2), >(1,5), ...]` and B `[..., >(1,5), (1,2), ...]`, where `>` is the iterator's current position. This intersection was queried once, so the set of `intersectionValues` contains `(1,2)`, and both A and B are in `intersectionIterators`.
The iterator is forwarded to `(1,5)`. Both iterators are already pointing to `(1,5)`, so they don't move, but the deduplication info is lost. Next two times the iterator is queried, we get `(1,5)`, then `(1,2)`, despite the fact that `(1,2)` was already output before the `forward()` call.
I think a solution could be gating the two added lines behind an `isIntersection(candidate, target)` check.

Reply:
Upon reflection, the different order isn't required either. `[..., (1,2), >(1,5), ...]` and `[..., >(1,2), (1,5), ...]` would spell the same issue. In fact, we could be in a  `[..., (1,2), (1,5), >...]`, `[..., >(1,7) ...]`  state, in which case the first line of this method would be an error.